src/Microsoft.Framework.Primitives/StringValues.cs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Framework.Primitives
+{
+    /// <summary>
+    /// Represents zero/null, one, or many strings in an efficient way.
+    /// </summary>
+    public struct StringValues : IList<string>
+    {
+        private static readonly string[] EmptyArray = new string[0];
+        public static readonly StringValues Empty = new StringValues(EmptyArray);
+
+        private readonly string _value;
+        private readonly string[] _values;
+
+        public StringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public StringValues(string[] values)
+        {
+            _value = null;
+            _values = values;
+        }
+
+        public static implicit operator StringValues(string value)
+        {
+            return new StringValues(value);
+        }
+
+        public static implicit operator StringValues(string[] values)
+        {
+            return new StringValues(values);
+        }
+
+        public static implicit operator string (StringValues values)
+        {
+            return values.GetStringValue();
+        }
+
+        public static implicit operator string[] (StringValues value)
+        {
+            return value.GetArrayValue();
+        }
+
+        public int Count => _values?.Length ?? (_value != null ? 1 : 0);
+
+        bool ICollection<string>.IsReadOnly
+        {
+            get { return true; }
+        }
+
+        string IList<string>.this[int index]
+        {
+            get { return this[index]; }
+            set { throw new NotSupportedException(); }
+        }
+
+        public string this[int key]
+        {
+            get
+            {
+                if (_values != null)
+                {
+                    return _values[key]; // may throw
+                }
+                if (key == 0 && _value != null)
+                {
+                    return _value;
+                }
+                return EmptyArray[0]; // throws
+            }
+        }
+
+        public override string ToString()
+        {
+            return GetStringValue() ?? string.Empty;
+        }
+
+        private string GetStringValue()
+        {
+            if (_values == null)
+            {
+                return _value;
+            }
+            switch (_values.Length)
+            {
+                case 0: return null;
+                case 1: return _values[0];
+                default: return string.Join(",", _values);
+            }
+        }
+
+        public string[] ToArray()
+        {
+            return GetArrayValue() ?? EmptyArray;
+        }
+
+        private string[] GetArrayValue()
+        {
+            if (_value != null)
```

Comment:
I dunno, I'd be really nervous about having one method with a different logical order to determine state. It's `if values !=null then values, else if value!=null then string[1]{value}, else string[0]`. See also Count property. A comment next to the fields would have been nice, sorry about that :/

Tbh that means the string ctor could really detect "null" and assign values to a `string[1]{null}` instead.

That said, it can be refined after the merge, so treat this like a followup remark and keep running with the shipit